File: leetcode20.py
#Leetcode #20 Valid Parentheses 

class Solution:
    def isValid(self, s: str) -> bool:
        if s is None:
            return False
        else:
            list1 = []
            index = 0
            for i in range(len(s)):
                if not list1:
                    list1.append(s[i])
                    index = index + 1
                elif list1[index - 1] is '{' and s[i] is '}':
                    list1.pop()
                    index = index - 1 
                elif list1[index - 1] is '[' and s[i] is ']': #리스트 네이밍 실수함
                    list1.pop()
                    index = index - 1
                elif list1[index - 1] is '(' and s[i] is ')':  # 리스트 네이밍 실수함 반성하자 ㅎㅎ
                    list1.pop()
                    index = index - 1
                else: 
                    list1.append(s[i])
                    index = index + 1

            if not list1:
                return True
            else:
                return False

File: leetcode20.java
class Solution {
    public boolean isValid(String s) {
        
        Stack<Character> stack = new Stack<>();
        boolean answer = true;
        
        for(int i = 0; i<s.length();i++)
        {
            if(s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '[')
            {
                stack.push(s.charAt(i));
            }
            
            if(s.charAt(i) == ')')
            {
                if(stack.empty() == true)
                {
                    return false;
                }
                if(stack.peek() == '(')
                {
                    stack.pop();
                }else
                {
                    answer = false;
                    break;
                }
            }else if(s.charAt(i) == '}')
            {
                if(stack.empty() == true )
                {
                    return false; 
                }
                else if(stack.peek() == '{')
                {
                    stack.pop();
                }else
                {
                    answer = false;
                    break;
                }
            }else if(s.charAt(i) == ']')
            {
                if(stack.empty() == true)
                {
                    return false;
                }
                if(stack.peek() == '[')
                {
                    stack.pop();
                }else
                {
                    answer = false;
                    break;
                }
            }
        }
        
        if(stack.size() >0)
        {
            answer = false;
        }
        
            return answer;
        
        
    }
}